core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -500,6 +501,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
         ]
 
 
+def validate_current_project_delivery_status_and_upcoming_comms_milestones(
+    workbook: dict[str, pd.DataFrame]
+) -> list["TownsFundRoundFourValidationFailure"] | None:
+    """
+    Validates the Project Progress table in the Round 4 submission workbook.
+
+    Checks for rows with a Project Delivery Status other than "4. Completed" where both
+    the Current Project Delivery Stage and Most Important Upcoming Comms Milestone are null.
+    If such invalid states are found, returns a list of TownsFundRoundFourValidationFailure instances.
+
+    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
+                     representing each sheet in the Round 4 submission.
+    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
+             otherwise returns None.
+    """
+    project_progress_df = workbook["Project Progress"]
+
+    project_incomplete_rows = project_progress_df[project_progress_df["Project Delivery Status"] != "4. Completed"]
+
+    na_values = {"", "< Select >", np.nan, None}
+
+    columns_to_check = [
+        "Current Project Delivery Stage",
+        "Most Important Upcoming Comms Milestone",
+        "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
+    ]
+
+    invalid_rows = pd.DataFrame()
+
+    for column in columns_to_check:
+        invalid_rows = pd.concat([invalid_rows, find_invalid_rows(project_incomplete_rows, column, na_values)], axis=0)
+
+    if len(invalid_rows) > 0:
+        return [
+            TownsFundRoundFourValidationFailure(
+                sheet="Project Progress",
+                section="Projects Progress Summary",
+                column=row["null_column"],
+                message=msgs.BLANK_IF_PROJECT_INCOMPLETE,
+                row_indexes=[row.name],
+            )
+            for _, row in invalid_rows.iterrows()
+        ]
+
+
+def find_invalid_rows(df: pd.DataFrame, column: str, na_values: set) -> pd.DataFrame:
+    """
+    Helper function to find rows with null values in a specified column.
+
+    Captures the name of the specified column in new column "null_column".
+    This enables retrieval of multiple columns that are non-nullable in the same row
+    if this function is called for each column to be validated.
+
+    :param df: DataFrame to search for invalid rows.
+    :param column: Column name to check for null values.
+    :param na_values: Set of values considered as null.
+    :return: DataFrame containing rows with null values in the specified column.
+    """
+    invalid_rows = df[df[column].isin(na_values)]
+    if len(invalid_rows) > 0:
+        invalid_rows["null_column"] = column
+    return invalid_rows
```

Comment:
I would be tempted to re-factor this so to avoid altering the dataframe rows.
```
invalid_rows =[]
for row in invalid_rows:
	for column in columns_to_check:
		if row[column].isin(na_values):
			invalid_rows.append(row.name, column)

return invalid_rows
```

Reply:
potentially less readably though

Reply:
We need the tacked-on ```"null_column"``` for our specific use case in which there are three columns we're assessing, but one or more of them could be wrong on the same line. As each validation failure needs to return the specific column where the error has occurred, we need a way of distinguishing from which column the failure has occurred if there are errors in multiple columns in one row. This is a workaround as the alternative seemed to be three separate DataFrames over which we iterate three times.

With that in mind, I don't think the above code woks?